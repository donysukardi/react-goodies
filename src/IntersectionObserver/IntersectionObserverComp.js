import { Component } from 'react'
import PropTypes from 'prop-types'

class IntersectionObserverComp extends Component {
  state = {}

  setRef = ref => {
    this.domNode = ref
  }

  onObserve = entries => {
    const { thresholds } = this.props
    this.setState(
      Object.keys(thresholds).reduce(
        (totalResult, prop) => ({
          ...totalResult,
          ...entries.reduce(
            (entriesResult, entry) => ({
              ...entriesResult,
              [prop]:
                entry.isIntersecting &&
                entry.intersectionRatio >= thresholds[prop],
            }),
            {},
          ),
        }),
        {},
      ),
    )
  }

  componentDidMount() {
    const { thresholds, options } = this.props
    this.observer = new global.IntersectionObserver(this.onObserve, {
      ...options,
      threshold: Object.keys(thresholds).map(prop => thresholds[prop]),
    })
    this.observer.observe(this.domNode)
  }

  componentWillUnmount() {
    this.observer.disconnect()
  }

  render() {
    const { children } = this.props
    return children({
      setRef: this.setRef,
      ...this.state,
    })
  }
}

IntersectionObserverComp.propTypes = {
  children: PropTypes.func.isRequired,
  options: PropTypes.object,
  thresholds: PropTypes.object,
}

IntersectionObserverComp.defaultProps = {
  thresholds: {},
  options: {},
}

export default IntersectionObserverComp
