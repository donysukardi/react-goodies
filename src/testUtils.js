const propsToDataAttrs = props =>
  Object.entries(props).reduce((acc, [key, value]) => {
    acc[`data-${String(key).toLowerCase()}`] = value
    return acc
  }, {})

export {propsToDataAttrs}
