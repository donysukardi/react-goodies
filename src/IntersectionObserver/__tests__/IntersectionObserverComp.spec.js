import React from 'react'
import PropTypes from 'prop-types'
import { cleanup, render } from 'react-testing-library'
import IntersectionObserverComp from '../IntersectionObserverComp'
import { propsToDataAttrs } from '../../testUtils'

const Target = ({ setRef, children, ...props }) => {
  setRef('ref')
  return (
    <div data-testid="targetContainer" {...propsToDataAttrs(props)}>
      {children}
    </div>
  )
}
Target.propTypes = {
  setRef: PropTypes.func.isRequired,
  children: PropTypes.node,
}

describe('withIntersectionObserverProps', () => {
  describe('Intersection Observer API is supported', () => {
    let origIntersectionObserver = null

    beforeAll(() => {
      origIntersectionObserver = global.IntersectionObserver
    })

    beforeEach(() => {
      jest.resetModules()

      global.IntersectionObserver = jest.fn(() => ({
        observe: () => {},
        disconnect: () => {},
      }))
    })

    afterEach(cleanup)

    afterAll(() => {
      global.IntersectionObserver = origIntersectionObserver
    })

    it('should register observer with DOM node from `setRef`', () => {
      const mockObserve = jest.fn()

      global.IntersectionObserver = jest.fn(() => ({
        observe: mockObserve,
      }))

      render(
        <IntersectionObserverComp
          thresholds={{
            foo: 0.5,
            bar: 1,
          }}
        >
          {props => <Target {...props} />}
        </IntersectionObserverComp>,
      )

      expect(global.IntersectionObserver.mock.calls).toMatchSnapshot()
      expect(mockObserve.mock.calls).toMatchSnapshot()
    })

    it('should unregister observer with DOM node on unmount', () => {
      const mockObserve = jest.fn()
      const mockUnobserve = jest.fn()

      global.IntersectionObserver = jest.fn(() => ({
        observe: mockObserve,
        disconnect: mockUnobserve,
      }))

      const { unmount } = render(
        <IntersectionObserverComp
          thresholds={{
            foo: 0.5,
            bar: 1,
          }}
        >
          {props => <Target {...props} />}
        </IntersectionObserverComp>,
      )

      unmount()

      expect(mockUnobserve.mock.calls).toMatchSnapshot()
    })

    it('should pass Intersection Observer options', () => {
      global.IntersectionObserver = jest.fn(() => ({
        observe: () => {},
      }))

      render(
        <IntersectionObserverComp
          options={{
            foo: 'bar',
          }}
        >
          {props => <Target {...props} />}
        </IntersectionObserverComp>,
      )

      expect(global.IntersectionObserver.mock.calls).toMatchSnapshot()
    })

    it('should map observer state to boolean props', () => {
      let observerCallback = null

      global.IntersectionObserver = jest.fn(callback => {
        observerCallback = callback

        return {
          observe: () => {},
        }
      })

      const { getByTestId } = render(
        <IntersectionObserverComp
          thresholds={{
            foo: 0.5,
            bar: 1,
          }}
        >
          {props => <Target {...props} />}
        </IntersectionObserverComp>,
      )
      expect(getByTestId('targetContainer')).toMatchSnapshot()

      observerCallback([{ isIntersecting: false, intersectionRatio: 0.5 }])
      expect(getByTestId('targetContainer')).toMatchSnapshot()

      observerCallback([{ isIntersecting: true, intersectionRatio: 0.5 }])
      expect(getByTestId('targetContainer')).toMatchSnapshot()

      observerCallback([{ isIntersecting: true, intersectionRatio: 1 }])
      expect(getByTestId('targetContainer')).toMatchSnapshot()
    })
  })
})
