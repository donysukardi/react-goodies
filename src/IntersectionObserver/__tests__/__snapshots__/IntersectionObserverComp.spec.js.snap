// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withIntersectionObserverProps Intersection Observer API is supported should map observer state to boolean props 1`] = `
<div
  data-testid="targetContainer"
/>
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should map observer state to boolean props 2`] = `
<div
  data-bar="false"
  data-foo="false"
  data-testid="targetContainer"
/>
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should map observer state to boolean props 3`] = `
<div
  data-bar="false"
  data-foo="true"
  data-testid="targetContainer"
/>
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should map observer state to boolean props 4`] = `
<div
  data-bar="true"
  data-foo="true"
  data-testid="targetContainer"
/>
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should pass Intersection Observer options 1`] = `
Array [
  Array [
    [Function],
    Object {
      "foo": "bar",
      "threshold": Array [],
    },
  ],
]
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should register observer with DOM node from \`setRef\` 1`] = `
Array [
  Array [
    [Function],
    Object {
      "threshold": Array [
        0.5,
        1,
      ],
    },
  ],
]
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should register observer with DOM node from \`setRef\` 2`] = `
Array [
  Array [
    "ref",
  ],
]
`;

exports[`withIntersectionObserverProps Intersection Observer API is supported should unregister observer with DOM node on unmount 1`] = `
Array [
  Array [],
]
`;
